#' doFlyMineAnalysis
#'
#' This function determines the lethality status in D melanogaster of a given human gene
#'
#' @param con A \code{SQLiteConnection} object
#' @param genes A vector of ENSEMBL gene ids
#' @return a data frame
#' @export
#'
#' @import biomaRt dplyr
doFlyMineAnalysis <- function(con, genes) {

    #get the fly orthologs
    fly_orthologs <- getOrthologs(con, genes, 'dmelanogaster')

    #group the fly orthologs for those genes with orthologs
    fly_orthologs_grouped <- fly_orthologs %>%
        dplyr::filter(nchar(dmelanogaster_homolog_orthology_type) > 4) %>%
        dplyr::group_by(ensembl_gene_id, dmelanogaster_homolog_orthology_type) %>%
        dplyr::summarise(dmelanogaster_count_orthologs=n(),
                  dmelanogaster_ortholog_ids=paste(dmelanogaster_homolog_ensembl_gene, collapse='|')) %>%
        dplyr::ungroup()

    #sort out entries with no ortholog
    no_fly_orthologs <-  fly_orthologs %>%
        dplyr::filter(nchar(dmelanogaster_homolog_orthology_type) < 4) %>%
        dplyr::transmute(ensembl_gene_id,
                         dmelanogaster_homolog_orthology_type='no_fly_ortholog',
                         dmelanogaster_count_orthologs=0,
                         dmelanogaster_ortholog_ids='')

    #combine and mark up the human genes with a single fly ortholog
    fly_orthologs_combined <- dplyr::bind_rows(fly_orthologs_grouped, no_fly_orthologs) %>%
        dplyr::mutate(single_ortholog_filter=(dmelanogaster_count_orthologs==1))

    #now format gene text for ruby script
    fly_gene_ids <- fly_orthologs_combined %>%
        dplyr::filter(single_ortholog_filter) %>%
        dplyr::transmute(id=dmelanogaster_ortholog_ids) %>%
        dplyr::distinct()

    fly_genes_txt <- paste(fly_gene_ids$id, collapse=',')

    #run the ruby script
    message(sprintf("Get data for %s fly genes from FlyMine", nrow(fly_gene_ids)))
    flymine_cmd <- sprintf('ruby \'%s/ruby/doFlyMineAnalysis.rb\' "%s"',system.file(package='CollateralVulnerability2016'), fly_genes_txt)
    flymine_data_txt <- system(flymine_cmd, intern=TRUE)
    flymine_data <- read.table(text=flymine_data_txt,sep="\t",header=F,skip=1, stringsAsFactors=F)

    message(sprintf('%s records retrieved, now writing to database', nrow(flymine_data)))
    #DBI::dbWriteTable(con, 'xxx', xxx, overwrite=T)
    message('Finished')

    return(flymine_data)




    return(martdata)
}
