#' getOrthologs
#'
#' This function uses the \pkg{biomaRt} package to get ortholog information for supplied human genes
#'
#' @param con A \code{SQLiteConnection} object
#' @param genes A vector of ENSEMBL gene ids
#' @param species The ENSEMBL species id eg mmusculus for mouse
#' @return a data frame
#' @export
#'
#' @import biomaRt dplyr
getOrthologs <- function(con, genes, species='mmusculus') {

    message(sprintf("Get the %s orthologs for %s human genes from biomart", species, length(genes)))

    #get the data from ensembl
    ensembl <- useMart("ENSEMBL_MART_ENSEMBL",dataset="hsapiens_gene_ensembl", host="mar2015.archive.ensembl.org")

    #check that the species is valid
    if (!(sprintf('%s_homolog_ensembl_gene', species) %in% listAttributes(ensembl)$name)) {
        stop(sprintf('%s not a valid ENSEMBL species id'))
    }

    #get the data from ensembl
    martdata <- getBM ( attributes = c('ensembl_gene_id',
                                       sprintf('%s_homolog_ensembl_gene', species),
                                       sprintf('%s_homolog_orthology_type', species),
                                       sprintf('%s_homolog_perc_id', species),
                                       sprintf('%s_homolog_perc_id_r1', species)),
                        filters = 'ensembl_gene_id',
                        values = genes,
                        mart = ensembl,
                        bmHeader=F)

    message('Done, now writing to database')
    DBI::dbWriteTable(con, sprintf("%s_martdata", species), martdata, overwrite=T)

    message('Finished')

    return(martdata)
}
