#' Flymine details table
#'
#' Extracts flymine analysis details from the SQLite database for a given gene
#'
#' @param con A \code{SQLiteConnection} object
#' @param input_gene An ensembl gene ID
#' @return A ggplot2 object
#' @import ggplot2
#' @export
tableFlymineDetails <- function(con, input_gene) {

    sql <- sprintf("
                    select
                    t2.primaryIdentifier,
                   t2.symbol,
                   t2.alleles_primaryIdentifier,
                   t2.alleles_symbol,
                   t2.alleles_phenotypeAnnotations_description as phenotypeDesc,
                   t2.is_lethal_fly
                   from
                   dmelanogaster_martdata t1
                   inner join flymine_analysis_flymine_filtered t2 on t1.dmelanogaster_homolog_ensembl_gene = t2.primaryIdentifier
                   where t1.ensembl_gene_id = '%s'
        ", input_gene)

    DBI::dbGetQuery(con, sql) %>%
        dplyr::mutate(FlybaseID = sprintf('<a href="http://www.flybase.org/%s target="_blank">%s</a>',
                                          primaryIdentifier, primaryIdentifier)) %>%
        as.data.frame()

}
